source/fundamentals/gridfs.txt

```diff
@@ -0,0 +1,387 @@
+
+When storing files with GridFS, the driver splits the files into smaller
+chunks, each represented by a separate document in the ``chunks`` collection.
+It also creates a document in the ``files`` collection that contains
+a file id as an ``ObjectId``, file name, and other file metadata. You can
+upload the file from memory or from a stream. See the following diagram to
+see how GridFS splits the files when uploaded to a bucket.
+
+.. figure:: /includes/figures/GridFS-upload.png
+   :alt: A diagram that shows how GridFS uploads a file to a bucket
+
+When retrieving files, GridFS fetches the metadata from the ``files``
+collection in the specified bucket and uses the information to reconstruct
+the file from documents in the ``chunks`` collection. You can read the file
+into memory or output it to a stream.
+
+.. _gridfs-create-bucket:
+
+Create a GridFS Bucket
+----------------------
+
+To use GridFS, first create a bucket or reference an existing one on
+a MongoDB database. Call the ``GridFSBuckets.create()`` helper method with
+a database instance as the parameter to instantiate a ``GridFSBucket``.
+You can use the ``GridFSBucket`` instance to call read and write operations on
+the files in your bucket.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start createGridFSBucket
+   :end-before: end createGridFSBucket
+
+.. _gridfs-create-custom-bucket:
+
+To create or reference a bucket with a custom name other than the default name
+``fs``, pass your bucket name as the second parameter to the ``create()``
+method as shown below:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start createCustomGridFSBucket
+   :end-before: end createCustomGridFSBucket
+
+For more information on the classes and methods mentioned in this section,
+see the following API documentation:
+
+- :java-docs:`GridFSBuckets <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSBuckets.html>`
+- :java-docs:`GridFSBucket <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSBucket.html>`
+
+.. _gridfs-store-files:
+
+Store Files
+-----------
+
+To store a file in a GridFS bucket, you can either upload it from an instance
+of ``InputStream`` or write its data to a ``GridFSUploadStream``.
+
+For either upload process, you can specify configuration information such
+as file chunk size and other field/value pairs to store as metadata. Set
+this information on an instance of ``GridFSUploadOptions`` as shown in the
+following code snippet:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start uploadOptions
+   :end-before: end uploadOptions
+
+See the :java-docs:`GridFSUploadOptions <apidocs/mongodb-driver-core/com/mongodb/client/gridfs/model/GridFSUploadOptions.html>`
+API documentation for more information.
```

Comment:
I wonder if we should try to give any guidance as to which one of these is most performance, robust, debuggable, extensible, etc. -- if I'm a user reading this page, it's helpful to see both options, but I would probably like a recommendation of which I should use as a sane default as well in case it looks like both would work for my use case.

Reply:
Are you asking whether Input or Output streams are better for performance/robustness/etc.? I think those are just two different ways to move data, and performance/robustness/etc. depends more on the Java built-in `io` package implementation.

Reply:
Fair enough -- if there really isn't much difference in usability or anything, seems fair enough to leave this as is with no explicit recommended default.